### DISCUSSION TITLE
Use the library directly without a client

### DISCUSSION THREAD

**Discussion by siddhsql** [qdrant/qdrant#3144] [General]
I want to ingest a large amount of data. Is there any way I can use the API that the server uses to access the database directly rather than having to go through a client?

**Comment by timvisee** [comment]
Yes, that's possible, though I'm not sure if it would be any faster.

More specifically, you could use the [gRPC](https://github.com/qdrant/qdrant/tree/master/lib/api/src/grpc/proto) (or [REST](https://qdrant.github.io/qdrant/redoc/index.html)) API that the server exposes directly.

Our Python client may not be the fastest, partially due to the language. If that is the case you may give our [Rust](https://github.com/qdrant/rust-client) client a shot.

Finally, I'd like to note that you should use our [batch](https://qdrant.tech/documentation/concepts/points/#upload-points) APIs for uploading a batch of points. That is much faster than upsering every point individually. All our clients have batching functions available.
